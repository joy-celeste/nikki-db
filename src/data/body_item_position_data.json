{"position":{"30":{"pot_scale":1,"posy":62,"posx":-7.5},"24":{"pot_scale":1,"posy":62,"posx":-9.5},"23":{"pot_scale":1,"posy":257.5,"posx":-18},"22":{"pot_scale":1,"posy":62,"posx":-7.5},"17":{"pot_scale":1,"posy":254,"posx":-23},"25":{"pot_scale":1,"posy":254,"posx":-23},"6":{"posy":299.5,"posx":-23.5},"7":{"posy":376,"posx":-30.5},"8":{"posy":260,"posx":-24},"9":{"posy":146,"posx":-22.5},"29":{"pot_scale":1,"posy":258,"posx":-12.5},"28":{"pot_scale":1,"posy":66.5,"posx":-11},"27":{"pot_scale":1,"posy":238.5,"posx":-26},"26":{"pot_scale":1,"posy":73.5,"posx":-7},"16":{"pot_scale":1,"posy":62,"posx":-9.5},"19":{"pot_scale":1,"posy":238.5,"posx":-26},"18":{"pot_scale":1,"posy":73.5,"posx":-7},"1":{"posy":-29.5,"posx":-22.5},"2":{"posy":239.5,"posx":-11.5},"3":{"posy":128.5,"posx":-21},"4":{"posy":253.5,"posx":-11.5},"5":{"posy":56,"posx":-9.5},"13":{"posy":253.5,"posx":-11.5},"12":{"posy":205.5,"posx":-24},"15":{"pot_scale":1,"posy":257.5,"posx":-18},"14":{"posy":56,"posx":-9.5},"10":{"posy":-204.5,"posx":-9.5},"11":{"posy":417.5,"posx":-21.5},"21":{"pot_scale":1,"posy":258,"posx":-12.5},"20":{"pot_scale":1,"posy":66.5,"posx":-11}},"id":1}