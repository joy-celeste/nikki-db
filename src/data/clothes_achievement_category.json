{"8":"Troupe","9":"4 Seasons","5":"Others","4":"Styling Contest","7":"Festivals","6":"7 Nations","3":"Journey","21":"Museum","20":"御苑琼芳","23":"Constellation Hall","22":"Ancient Fossil Hall","15":"North","14":"Pigeon","17":"Ruin","16":"Wasteland","18":"Happiness","19":"Story Suit","2":"Collection","1":"My Growth","12":"Lilith","13":"Cloud","10":"Stars","11":"Apple"}