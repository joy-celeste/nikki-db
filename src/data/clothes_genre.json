{"Coat":3,"AccessoryCount":8,"AccessoryEnd":17,"Earring":11,"Shoes":7,"Necklace":12,"Count":10,"Top":4,"Special":16,"Bottom":5,"Hair":1,"Dress":2,"Headwear":10,"Accessory":8,"AccessoryStart":10,"Waistdeco":15,"Bracelet":13,"Socks":6,"Skin":17,"MiracleSoul":88,"Makeup":9,"Handheld":14}