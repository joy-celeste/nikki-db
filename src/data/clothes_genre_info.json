{"1":{"combat_param":0.5,"display_name":"Hair","name":"发型","sub_type":[1],"file_name":"faxing"},"2":{"combat_param":2,"display_name":"Dress","name":"连衣裙","sub_type":[2],"file_name":"lianyiqun"},"3":{"combat_param":0.2,"display_name":"Coat","name":"外套","sub_type":[3],"file_name":"waitao"},"4":{"combat_param":1,"display_name":"Top","name":"上衣","sub_type":[4],"file_name":"shangyi"},"5":{"combat_param":1,"display_name":"Bottom","name":"下装","sub_type":[5],"file_name":"xiazhuang"},"6":{"combat_param":0.3,"display_name":"Hosiery","name":"袜子","sub_type":[6,7],"file_name":"wazi"},"7":{"combat_param":0.4,"display_name":"Shoes","name":"鞋子","sub_type":[8],"file_name":"xiezi"},"8":{"display_name":"Accessory","name":"饰品","file_name":"shipin","combat_param":0.2},"9":{"combat_param":0.1,"display_name":"Makeup","name":"妆容","sub_type":[9],"file_name":"zhuangrong"},"10":{"combat_param":0.2,"display_name":"Headwear","name":"头饰","sub_type":[10,20,28,29],"file_name":"toushi"},"11":{"combat_param":0.2,"display_name":"Earrings","name":"耳饰","sub_type":[11],"file_name":"ershi"},"12":{"combat_param":0.2,"display_name":"Necklace","name":"颈饰","sub_type":[12,13],"file_name":"jingshi"},"13":{"combat_param":0.2,"display_name":"Hands","name":"手饰","sub_type":[14,15,16],"file_name":"shoushi"},"14":{"combat_param":0.2,"display_name":"Handheld","name":"手持物","sub_type":[17,18,33],"file_name":"shouchiwu"},"15":{"combat_param":0.2,"display_name":"Waist","name":"腰饰","sub_type":[19],"file_name":"yaoshi"},"16":{"combat_param":0.2,"display_name":"Special","name":"特殊","sub_type":[21,22,23,24,25,26,27,30,31,35],"file_name":"teshu"},"17":{"combat_param":0.2,"display_name":"Skin","name":"皮肤","sub_type":[32],"file_name":"pifu"},"88":{"combat_param":0.2,"display_name":"Spirit","name":"萤光之灵","sub_type":[34],"file_name":"yingguangzhiling"}}