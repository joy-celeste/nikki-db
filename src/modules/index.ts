import { combineReducers } from 'redux';
import { userReducer } from './user';
import { productsReducer } from './products';

export const rootReducer = combineReducers({
  user: userReducer,
  products: productsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;