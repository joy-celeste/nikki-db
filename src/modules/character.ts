import { Dispatch } from 'react';
import { AnyAction } from 'redux';
import { RootState } from '.';
import { ACTION_CONSTANTS, DEFAULT_BODY, DEFAULT_CLOTHES, DEFAULT_AMPUTATIONS, SUBTYPES, BODY } from './constants';
import { AmputationData, AmputationParts, ItemData, ItemId, ItemsData, SubType } from './data';
import { Item } from './item';

export type BodyPart = number;
export type BodyParts = Set<BodyPart>;
export type Clothes = Record<SubType, ItemId>;
export type Amputations = Record<AmputationParts, ItemId[]>;

export class Character {
  visibleParts: BodyParts;
  clothes: Clothes;
  amputations: Amputations;

  constructor(input?: any) {
    this.visibleParts = input && input.visibleParts ? new Set(input.visibleParts) : new Set(DEFAULT_BODY);
    this.clothes = input && input.clothes ? { ...input.clothes } : { ...DEFAULT_CLOTHES };
    this.amputations = input && input.amputations ? { ...input.amputations } : { ...DEFAULT_AMPUTATIONS };
  }

  hide(bodyPart: BodyPart): void {
    if (this.visibleParts.has(bodyPart)) {
      this.visibleParts.delete(bodyPart);
    }
  }

  show(bodyPart: BodyPart): void {
    this.visibleParts.add(bodyPart);
  }

  wear(itemData: ItemData): void {
    const { subType, id, amputationData} = itemData;

    if (this.clothes[subType] && this.clothes[subType] === id && subType !== SUBTYPES.HAIR) {
      this.remove(subType);
      return;
    }

    if (this.clothes[subType] && this.clothes[subType] !== id) {
      this.remove(subType);
    }

    if (subType === SUBTYPES.DRESS) {
      this.remove(SUBTYPES.TOP);
      this.remove(SUBTYPES.BOTTOM);
    } else if (subType === SUBTYPES.TOP) {
      this.remove(SUBTYPES.DRESS);
    } else if (subType === SUBTYPES.BOTTOM) {
      this.remove(SUBTYPES.DRESS);
    }

    this.clothes[subType] = id;
    this.updateAmputations(id, amputationData, subType);
    this.updateVisibleBodyParts();
  }

  remove(subtype: SubType): void {
    if (this.clothes[subtype]) {
      const foundItem : ItemId = this.clothes[subtype];
      delete this.clothes[subtype];

      Object.keys(this.amputations).forEach((bodyPartStr) => {
        const bodyPart = parseInt(bodyPartStr, 10) as AmputationParts;
        this.amputations[bodyPart] = this.amputations[bodyPart].filter((itemId: ItemId) => itemId !== foundItem);
        if (this.amputations[bodyPart].length === 0) {
          this.show(bodyPart);
        }
      });
    }

    if (subtype === SUBTYPES.DRESS) {
      this.show(BODY.BRA);
      this.show(BODY.PANTY);
    } else if (subtype === SUBTYPES.TOP) {
      this.show(BODY.BRA);
    } else if (subtype === SUBTYPES.BOTTOM) {
      this.show(BODY.PANTY);
    }
  }

  updateAmputations(itemId: ItemId, amputationData: AmputationData, subtype: SubType): void {
    Object.keys(this.amputations).forEach((bodyPartStr) => {
      const bodyPart = parseInt(bodyPartStr, 10) as AmputationParts;
      if (this.shouldHideBodyPart(bodyPart, amputationData, subtype)) {
        this.amputations[bodyPart] = [...this.amputations[bodyPart], itemId];
      }
    });
  }

  shouldHideBodyPart(bodyPart: BodyPart, amputationData: AmputationData, subtype: SubType): boolean {
    if (amputationData && amputationData[bodyPart]) {
      return true;
    }

    if ((bodyPart === BODY.BRA || bodyPart === BODY.VEST) && (subtype === SUBTYPES.TOP || subtype === SUBTYPES.DRESS)) {
      return true;
    }

    if (bodyPart === BODY.PANTY && (subtype === SUBTYPES.BOTTOM || subtype === SUBTYPES.DRESS)) {
      return true;
    }

    return false;
  }

  // also refactor wear to only take in itemdata instead of having to use 3 params

  updateVisibleBodyParts(): void {
    // If any amputation data is associated with an amputation location, hide the limb
    // Else if nothing is being worn on this amputation location right now, make it visible
    Object.entries(this.amputations).forEach(([bodyPartStr, items]) => {
      const bodyPart = parseInt(bodyPartStr, 10) as BodyPart;
      if (items.length >= 1) {
        this.hide(bodyPart); // have to cast as int as Object.keys turns keys into string
      } else {
        this.show(bodyPart);
      }
    });
  }
}

export type CharacterState = {
  history: Character[];
  step: number;
};

const initialState: CharacterState = {
  history: [new Character()],
  step: 0,
};

// ACTIONS
const addToHistory = (character: Character, step: number): AnyAction => ({
  type: ACTION_CONSTANTS.CHARACTER_ADD_TO_HISTORY,
  payload: {
    character, step,
  },
});

// USE-CASE
export const wearItem = (itemId: ItemId) =>
  async(dispatch: Dispatch<AnyAction>, getState: () => RootState): Promise<any> => {
    const itemDataState: ItemsData = getState().data.itemsData;
    const itemData = itemDataState[itemId];

    if (itemData && itemData.subType) {
      const charState: CharacterState = getState().character;
      const oldChar: Character = charState.history[charState.step];
      const newChar: Character = new Character(oldChar);
      newChar.wear(itemData);
      dispatch(addToHistory(newChar, charState.step + 1));
    }
  };

// REDUCER
export function characterReducer(
  state = initialState,
  action: AnyAction,
): CharacterState {
  switch (action.type) {
    case ACTION_CONSTANTS.CHARACTER_ADD_TO_HISTORY:
      return {
        ...state,
        history: [...state.history, action.payload.character],
        step: action.payload.step,
      };
    default:
      return state;
  }
}
