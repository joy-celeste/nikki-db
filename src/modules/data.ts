import { AnyAction } from 'redux';
import { DocumentData } from '@firebase/firestore-types';
import { DeserializeNullException } from './errors';
import { ACTION_CONSTANTS, BODY, BODY_ITEM_DATA, BODY_ITEM_ID, BODY_PARTS_DEPTHS,
  DEFAULT_AMPUTATIONS_LIST, DEPTHTYPE_TO_SUBTYPES, SUBTYPES_LIST, CLOTHES_DATA } from './constants';
import { Character, CharacterState, wearItem, addToHistory } from './character';
import { RootState } from '.';
import { changeHiddenItemList } from './editor';

export type SubType = typeof SUBTYPES_LIST[number];
export type AmputationParts = typeof DEFAULT_AMPUTATIONS_LIST[number];
export type AmputationData = Record<AmputationParts, boolean>;
export type Depths = Record<number, number>;
export type ItemId = number;
export type DepthType = number;

export interface SubtypeInfo {
  subtype: number,
  depth: Record<number, number>
}

export class ItemData {
  id?: ItemId;
  name?: string;
  description?: string;
  depthType?: DepthType;
  subType?: SubType;
  amputationData?: AmputationData;
  position?: PositionData[];
  depths?: Depths;

  constructor(input?: DocumentData) {
    this.deserialize(input);
  }

  deserialize(input: DocumentData): ItemData {
    if (!input) {
      throw new DeserializeNullException('Cannot deserialize null input for ItemData');
    }

    this.id = input.id || null;
    this.name = input.name || '';
    this.description = input.desc || '';
    this.position = null;
    this.depthType = null;
    this.subType = null;
    this.depths = null;
    this.amputationData = null;

    if (input.amputation_data) {
      this.amputationData = {
        [BODY.ARM]: !!input.amputation_data.arm,
        [BODY.TORSO]: !!input.amputation_data.body,
        [BODY.BREAST]: !!input.amputation_data.body,
        [BODY.LEG]: !!input.amputation_data.leg,
        [BODY.PANTY]: !!input.amputation_data.panty,
      };
    }

    if (input.depth_type || input.id === BODY_ITEM_ID) {
      this.depthType = input.depth_type;
      const subtypeData = input.id !== BODY_ITEM_ID
        ? DEPTHTYPE_TO_SUBTYPES[input.depth_type.toString()]
        : BODY_PARTS_DEPTHS;
      this.subType = subtypeData.subtype;
      this.depths = subtypeData.depth;
    }

    if (input.position) {
      const position: PositionData[] = [];
      Object.entries(input.position).forEach(([key, value]) => {
        position.push(new PositionData(parseInt(key, 10), value, this.depths));
      });
      this.position = position;
    }
    return this;
  }
}

export class PositionData {
  depth: number;
  x: number;
  y: number;
  z: number;
  height?: number;
  width?: number;
  scale?: number;

  constructor(depth: number, input: DocumentData, depths?: Depths) {
    this.deserialize(depth, input, depths);
  }

  deserialize(depth: number, input: DocumentData, depths?: Depths): PositionData {
    if (!depth || !input) {
      throw new DeserializeNullException('Cannot deserialize null input for PositionData');
    }

    this.depth = depth;
    this.x = input.posx;
    this.y = input.posy;
    this.z = null;
    this.height = input.height || null;
    this.width = input.width || null;
    this.scale = input.pot_scale || null;

    if (depths) {
      this.calculateDepth(depth, depths);
    }
    return this;
  }

  // Don't ask me about this logic... I'm just copying this straight from the original devs
  calculateDepth(depth: number, depths: Depths): void {
    if (depth < 100) {
      this.z = depths[depth];
    } else {
      const key: number = depth / 100;
      const ceil: number = Math.ceil(key);
      /* istanbul ignore next */
      if (key <= 0 || ceil === key) { // These are for animated effects.
        this.z = depths[ceil] - 5;
      }
      this.z = depths[ceil - 1] - 5;
    }
  }
}

export const bodyItemData = new ItemData(BODY_ITEM_DATA);

export type ItemsData = Record<ItemId, ItemData>;
export type DataState = {
  itemsData: ItemsData;
  loading: boolean;
};

export const initialState: DataState = {
  itemsData: { 10001: new ItemData(CLOTHES_DATA[10001]) },
  loading: false,
};

// ACTIONS
export const addItemData = (itemId: number, itemData: ItemData): AnyAction => ({
  type: ACTION_CONSTANTS.DATA_ADD_ITEMS,
  payload: {
    itemId, itemData,
  },
});

export type BreakdownData = {
  variant: string;
  itemId: ItemId;
  prosthesis: boolean;
  subtype: SubType;
  name: string;
};

// USE-CASE
export const loadItem = (itemId: ItemId) =>
  async(dispatch: Function, getState: () => RootState): Promise<void> => {
    const items: ItemsData = getState().data.itemsData;
    if (!(itemId in items) && CLOTHES_DATA[itemId]) {
      dispatch(addItemData(itemId, new ItemData(CLOTHES_DATA[itemId])));
    } else if (itemId in items) {
      dispatch(wearItem(itemId));
    }
  };

export const loadMultipleItems = (itemIds: ItemId[]) =>
  async(dispatch: Function, getState: () => RootState): Promise<void> => {
    dispatch(changeHiddenItemList(new Set<ItemId>()));
    if (itemIds.length === 1) {
      dispatch(loadItem(itemIds[0]));
      return;
    }

    const items: ItemsData = getState().data.itemsData;
    const charState: CharacterState = getState().character;
    const newChar: Character = new Character();
    let itemData: ItemData;

    await Promise.allSettled(itemIds.map(async (itemId) => {
      if (!(itemId in items)) {
        itemData = new ItemData(CLOTHES_DATA[itemId]);
        dispatch(addItemData(itemId, itemData));
      } else {
        itemData = items[itemId];
      }
      newChar.wear(itemData);
    })).then(() => {
      dispatch(addToHistory(newChar, charState.step + 1));
    });
  };

// REDUCER
export function dataReducer(
  state = initialState,
  action: AnyAction,
): DataState {
  switch (action.type) {
    case ACTION_CONSTANTS.DATA_ADD_ITEMS:
      return {
        ...state,
        itemsData: {
          ...state.itemsData,
          [action.payload.itemId]: action.payload.itemData,
        },
      };
    default:
      return state;
  }
}
