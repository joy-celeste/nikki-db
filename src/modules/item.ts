import { bodyItemData, ItemData, PositionData } from './data';
import { BodyParts } from './character';
import { NoDataException } from './errors';

export class Piece {
  isVisible?: boolean = true;
  depth: number;
  x: number;
  y: number;
  z?: number;

  constructor(position: PositionData, body?: BodyParts) {
    this.x = position.x;
    this.y = position.y;
    this.z = position.z;
    this.depth = position.depth;
    this.isVisible = body ? body.has(position.depth) : true;
  }
}

export class Item {
  itemId: number;
  pieces: Piece[];

  constructor(itemData: ItemData) {
    if (!itemData || !itemData.id || !itemData.position) {
      throw new NoDataException('Cannot construct new Item from no itemData!');
    }

    this.itemId = itemData.id;
    this.pieces = itemData.position.map((positionData: PositionData) => new Piece(positionData));
    return this;
  }
}

export class Body extends Item {
  pieces: Piece[];

  constructor(visibleParts: BodyParts) {
    super(bodyItemData);
    this.pieces = bodyItemData.position.map((positionData: PositionData) => new Piece(positionData, visibleParts));
  }
}
