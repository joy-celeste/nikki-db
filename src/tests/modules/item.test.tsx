import { posedCoat, posedDress } from '../test_data/data';
import { Body, Item } from '../../modules/item';
import { NoDataException } from '../../modules/errors';
import { ItemData } from '../../modules/data';
import { Character } from '../../modules/character';

describe('Item', () => {
  test('Returns an object with null data if given no ItemData', () => {
    expect(() => new Item(null)).toThrowError(NoDataException);
  });

  test('Returns an BODY when given visible parts data', () => {
    const character: Character = new Character();
    const body = new Body(character.visibleParts);
    expect(body).toMatchSnapshot();
  });

  test('Returns an ITEM object with pieces if given ItemData', () => {
    const coat = new ItemData(posedCoat);
    const coatItem = new Item(coat);
    expect(coatItem).toMatchSnapshot();
  });

  test('Returns an ITEM object with pieces if given ItemData', () => {
    const dress = new ItemData(posedDress);
    const dressItem = new Item(dress);
    expect(dressItem).toMatchSnapshot();
  });
});
