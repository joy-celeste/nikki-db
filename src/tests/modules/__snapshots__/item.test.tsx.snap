// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Returns an BODY when given visible parts data 1`] = `
Body {
  "itemId": 1,
  "pieces": Array [
    Piece {
      "depth": 1,
      "isVisible": false,
      "x": -22.5,
      "y": -29.5,
      "z": 10000,
    },
    Piece {
      "depth": 2,
      "isVisible": true,
      "x": -11.5,
      "y": 239.5,
      "z": 10020,
    },
    Piece {
      "depth": 3,
      "isVisible": false,
      "x": -21,
      "y": 128.5,
      "z": 24000,
    },
    Piece {
      "depth": 4,
      "isVisible": true,
      "x": -11.5,
      "y": 253.5,
      "z": 10040,
    },
    Piece {
      "depth": 5,
      "isVisible": true,
      "x": -9.5,
      "y": 56,
      "z": 10030,
    },
    Piece {
      "depth": 8,
      "isVisible": false,
      "x": -24,
      "y": 260,
      "z": 10000,
    },
    Piece {
      "depth": 9,
      "isVisible": true,
      "x": -22.5,
      "y": 146,
      "z": 10000,
    },
    Piece {
      "depth": 10,
      "isVisible": true,
      "x": -9.5,
      "y": -204.5,
      "z": 10000,
    },
    Piece {
      "depth": 11,
      "isVisible": true,
      "x": -21.5,
      "y": 417.5,
      "z": 10000,
    },
    Piece {
      "depth": 12,
      "isVisible": true,
      "x": -24,
      "y": 205.5,
      "z": 10000,
    },
    Piece {
      "depth": 13,
      "isVisible": false,
      "x": -11.5,
      "y": 253.5,
      "z": 10400,
    },
    Piece {
      "depth": 14,
      "isVisible": false,
      "x": -9.5,
      "y": 56,
      "z": 10400,
    },
    Piece {
      "depth": 15,
      "isVisible": false,
      "x": -18,
      "y": 257.5,
      "z": 10040,
    },
    Piece {
      "depth": 16,
      "isVisible": false,
      "x": -9.5,
      "y": 62,
      "z": 10030,
    },
    Piece {
      "depth": 17,
      "isVisible": false,
      "x": -23,
      "y": 254,
      "z": 10040,
    },
    Piece {
      "depth": 18,
      "isVisible": false,
      "x": -7,
      "y": 73.5,
      "z": 10030,
    },
    Piece {
      "depth": 19,
      "isVisible": false,
      "x": -26,
      "y": 238.5,
      "z": 10040,
    },
    Piece {
      "depth": 20,
      "isVisible": false,
      "x": -11,
      "y": 66.5,
      "z": 10030,
    },
    Piece {
      "depth": 21,
      "isVisible": false,
      "x": -12.5,
      "y": 258,
      "z": 10040,
    },
    Piece {
      "depth": 22,
      "isVisible": false,
      "x": -7.5,
      "y": 62,
      "z": 10030,
    },
    Piece {
      "depth": 23,
      "isVisible": false,
      "x": -18,
      "y": 257.5,
      "z": undefined,
    },
    Piece {
      "depth": 24,
      "isVisible": false,
      "x": -9.5,
      "y": 62,
      "z": undefined,
    },
    Piece {
      "depth": 25,
      "isVisible": false,
      "x": -23,
      "y": 254,
      "z": undefined,
    },
    Piece {
      "depth": 26,
      "isVisible": false,
      "x": -7,
      "y": 73.5,
      "z": undefined,
    },
    Piece {
      "depth": 27,
      "isVisible": false,
      "x": -26,
      "y": 238.5,
      "z": undefined,
    },
    Piece {
      "depth": 28,
      "isVisible": false,
      "x": -11,
      "y": 66.5,
      "z": undefined,
    },
    Piece {
      "depth": 29,
      "isVisible": false,
      "x": -12.5,
      "y": 258,
      "z": undefined,
    },
    Piece {
      "depth": 30,
      "isVisible": false,
      "x": -7.5,
      "y": 62,
      "z": undefined,
    },
  ],
}
`;

exports[`Item Returns an ITEM object with pieces if given ItemData 1`] = `
Item {
  "itemId": 30987,
  "pieces": Array [
    Piece {
      "depth": 1,
      "isVisible": true,
      "x": -27,
      "y": 200.5,
      "z": 36700,
    },
    Piece {
      "depth": 2,
      "isVisible": true,
      "x": -202,
      "y": -130,
      "z": 27000,
    },
    Piece {
      "depth": 3,
      "isVisible": true,
      "x": 101,
      "y": -119,
      "z": 19000,
    },
    Piece {
      "depth": 4,
      "isVisible": true,
      "x": -81,
      "y": -153,
      "z": 5000,
    },
    Piece {
      "depth": 13,
      "isVisible": true,
      "x": -88.5,
      "y": 245,
      "z": 10160,
    },
    Piece {
      "depth": 14,
      "isVisible": true,
      "x": 42.5,
      "y": 234.5,
      "z": 9880,
    },
    Piece {
      "depth": 101,
      "isVisible": true,
      "x": -30.5,
      "y": 242.5,
      "z": 36695,
    },
    Piece {
      "depth": 201,
      "isVisible": true,
      "x": -47.5,
      "y": 262,
      "z": 26995,
    },
    Piece {
      "depth": 301,
      "isVisible": true,
      "x": 124.5,
      "y": 139,
      "z": 18995,
    },
  ],
}
`;

exports[`Item Returns an ITEM object with pieces if given ItemData 2`] = `
Item {
  "itemId": 22008,
  "pieces": Array [
    Piece {
      "depth": 2,
      "isVisible": true,
      "x": -32.5,
      "y": -164.5,
      "z": 24000,
    },
    Piece {
      "depth": 901,
      "isVisible": true,
      "x": -59.5,
      "y": 328,
      "z": 10215,
    },
  ],
}
`;
